File: Backend/Timer/CountdownTimer.cs
namespace Backend.Timer;

public sealed record TimerAnchor(bool Running, long AnchorTimestamp, double DurationLeftAtAnchor)
{
	public static TimerAnchor Idle { get; } = new(Running: false, AnchorTimestamp: 0, DurationLeftAtAnchor: 0);
}

public static class CountdownTimer
{
	public static TimerAnchor Start(double durationSeconds, long now) =>
		new(Running: true, AnchorTimestamp: now, DurationLeftAtAnchor: durationSeconds);

	public static double DeriveTimeLeft(TimerAnchor anchor, long now) =>
		anchor.Running
			? Math.Max(0, anchor.DurationLeftAtAnchor - ((now - anchor.AnchorTimestamp) / 1000.0))
			: anchor.DurationLeftAtAnchor;

	public static bool HasExpired(TimerAnchor anchor, long now) =>
		anchor.Running && DeriveTimeLeft(anchor, now) <= 0;
}

File: shared/core/countdownTimer.ts
// Pure countdown transforms — the TS mirror of C# `Core.Timer.CountdownTimer`.
// Durations are seconds (the C# `TimeSpan` collapses to its total seconds on the wire).

import type { TimerAnchor } from './types';

export function start(durationSeconds: number, now: number): TimerAnchor {
	return { running: true, anchorTimestamp: now, durationLeftAtAnchor: durationSeconds };
}

export function deriveTimeLeft(anchor: TimerAnchor, now: number): number {
	return anchor.running
		? Math.max(0, anchor.durationLeftAtAnchor - (now - anchor.anchorTimestamp) / 1000)
		: anchor.durationLeftAtAnchor;
}

export function hasExpired(anchor: TimerAnchor, now: number): boolean {
	return anchor.running && deriveTimeLeft(anchor, now) <= 0;
}

export function formatTimeLeft(seconds: number): string {
	const total = Math.ceil(seconds);
	const minutes = Math.floor(total / 60);
	const remainder = total % 60;
	return `${minutes}:${remainder.toString().padStart(2, '0')}`;
}
